public/app/features/dashboard-scene/saving/getDashboardChanges.ts

```diff
@@ -88,35 +109,107 @@ export function getRawDashboardChanges(
   };
 }
 
-export function getHasTimeChanged(saveModel: Dashboard, originalSaveModel: Dashboard) {
-  return saveModel.time?.from !== originalSaveModel.time?.from || saveModel.time?.to !== originalSaveModel.time?.to;
+interface DefaultPersistedTimeValue {
+  from?: string;
+  to?: string;
+}
+export function getHasTimeChanged(
+  newRange: DefaultPersistedTimeValue = {},
+  previousRange: DefaultPersistedTimeValue = {}
+) {
+  return newRange.from !== previousRange.from || newRange.to !== previousRange.to;
 }
 
-export function adHocVariableFiltersEqual(a: AdHocVariableModel, b: AdHocVariableModel) {
-  if (a.filters === undefined && b.filters === undefined) {
+export function adHocVariableFiltersEqual(filtersA?: AdHocFilterWithLabels[], filtersB?: AdHocFilterWithLabels[]) {
+  if (filtersA === undefined && filtersB === undefined) {
     console.warn('Adhoc variable filter property is undefined');
     return true;
   }
 
-  if ((a.filters === undefined && b.filters !== undefined) || (b.filters === undefined && a.filters !== undefined)) {
+  if ((filtersA === undefined && filtersB !== undefined) || (filtersB === undefined && filtersA !== undefined)) {
     console.warn('Adhoc variable filter property is undefined');
     return false;
   }
 
-  if (a.filters.length !== b.filters.length) {
+  if (filtersA?.length !== filtersB?.length) {
     return false;
   }
 
-  for (let i = 0; i < a.filters.length; i++) {
-    const aFilter = a.filters[i];
-    const bFilter = b.filters[i];
-    if (aFilter.key !== bFilter.key || aFilter.operator !== bFilter.operator || aFilter.value !== bFilter.value) {
+  for (let i = 0; i < (filtersA?.length ?? 0); i++) {
+    const aFilter = filtersA?.[i];
+    const bFilter = filtersB?.[i];
+    if (aFilter?.key !== bFilter?.key || aFilter?.operator !== bFilter?.operator || aFilter?.value !== bFilter?.value) {
       return false;
     }
   }
   return true;
 }
 
+export function applyVariableChangesV2(
+  saveModel: DashboardV2Spec,
+  originalSaveModel: DashboardV2Spec,
+  saveVariables?: boolean
+) {
+  const originalVariables = originalSaveModel.variables ?? [];
+  const variablesToSave = saveModel.variables ?? [];
+  let hasVariableValueChanges = false;
+
+  for (const variable of variablesToSave) {
+    const hasCurrentValueToSave = (v: VariableKind) =>
+      v.kind === 'QueryVariable' ||
+      v.kind === 'CustomVariable' ||
+      v.kind === 'DatasourceVariable' ||
+      v.kind === 'ConstantVariable' ||
+      v.kind === 'IntervalVariable' ||
+      v.kind === 'TextVariable' ||
+      v.kind === 'GroupByVariable';
+
+    const hasOptionsToSave = (v: VariableKind) =>
+      v.kind === 'QueryVariable' ||
+      v.kind === 'CustomVariable' ||
+      v.kind === 'DatasourceVariable' ||
+      v.kind === 'IntervalVariable' ||
+      v.kind === 'GroupByVariable';
```

Comment:
Depending on the query type, we need to store options and current. TS helps to do this, but we just need this two type checkers to make sure we don't try to access to options when the variable doesn't have